src/bootstrap/src/utils/execution_context.rs

```diff
@@ -80,15 +81,16 @@ impl ExecutionContext {
     /// Note: Ideally, you should use one of the BootstrapCommand::run* functions to
     /// execute commands. They internally call this method.
     #[track_caller]
-    pub fn run(
+    pub fn start<'a>(
         &self,
-        command: &mut BootstrapCommand,
+        command: &'a mut BootstrapCommand,
         stdout: OutputMode,
         stderr: OutputMode,
-    ) -> CommandOutput {
+    ) -> DeferredCommand<'a> {
         command.mark_as_executed();
+
         if self.dry_run() && !command.run_always {
-            return CommandOutput::default();
+            return DeferredCommand::new(None, stdout, stderr, command, Arc::new(self.clone()));
```

Comment:
Let's just build the deferred command using struct creation literal (`DeferredCommand { stdout, stderr, ... }`), the `new` function has a bunch of parameters and it's hard to keep track what parameter is on which index. Also then we can remove `new` altogether, I think it shouldn't be public, it should only be possible to create deferred commands using the `start` method here, and nothing else.